src/sys/fuchsia/eventedfd.rs

```diff
@@ -0,0 +1,246 @@
+            inner: Arc::new(EventedFdInner {
+                registration: Mutex::new(None),
+                fd: fd,
+                mxio: mxio,
+            })
+        }
+    }
+}
+
+impl Evented for EventedFd {
+    fn register(&self,
+                poll: &Poll,
+                token: Token,
+                interest: Ready,
+                opts: PollOpt) -> io::Result<()>
+    {
+        let epoll_events = ioevent_to_epoll(interest, opts);
+
+        let (handle, raw_handle, signals) = unsafe {
+            let mut raw_handle: sys::mx_handle_t = mem::uninitialized();
+            let mut signals: sys::mx_signals_t = mem::uninitialized();
+            sys::__mxio_wait_begin(self.inner.mxio, epoll_events, &mut raw_handle, &mut signals);
+
+            // We don't have ownership of the handle, so we can't drop it
+            let handle = DontDrop::new(magenta::Handle::from_raw(raw_handle));
+            (handle, raw_handle, signals)
+        };
+
+
+        let needs_rereg = opts.is_level() && !opts.is_oneshot();
+
+        {
+            let mut registration_lock = self.inner.registration.lock().unwrap();
+            if registration_lock.is_some() {
+                return Err(io::Error::new(
+                    io::ErrorKind::AlreadyExists,
+                    "Called register on an already registered file descriptor."));
+            }
+            *registration_lock = Some(EventedFdRegistration {
+                token: token,
+                handle: DontDrop::new(unsafe { magenta::Handle::from_raw(raw_handle) }),
+                rereg_signals: if needs_rereg {
+                    Some((signals, poll_opts_to_wait_async(opts)))
+                } else {
+                    None
+                },
+            })
+        }
+
+        let registered = poll::selector(poll)
+            .register_fd(handle.inner_ref(), self, token, signals, opts);
+
+        if registered.is_err() {
+            let mut registration_lock = self.inner.registration.lock().unwrap();
+            *registration_lock = None;
+        }
+
+        registered
+    }
+
+    fn reregister(&self,
+                  poll: &Poll,
+                  token: Token,
+                  interest: Ready,
+                  opts: PollOpt) -> io::Result<()>
+    {
+        // Take out the registration lock
+        let mut registration_lock = self.inner.registration.lock().unwrap();
+
+        // Deregister
+        {
+            let old_registration = if let Some(old_reg) = registration_lock.take() {
+                old_reg
+            } else {
+                return Err(io::Error::new(
+                    io::ErrorKind::NotFound,
+                    "Called rereregister on an unregistered file descriptor."))
+            };
+
+            poll::selector(poll)
+                .deregister_fd(old_registration.handle.inner_ref(), old_registration.token)?;
+        }
+
+        // Reregister
+        let epoll_events = ioevent_to_epoll(interest, opts);
```

Comment:
Could this stuff that's duplicated with `register` be factored out into a helper that takes in the `MutexGuard` for the registration lock and the `self.inner.mxio`? Then both `register` and `reregister` could call it, just with different stuff around it.

Alternatively, the block that gets the handle and signals and the block that sets up the registration lock could be factored into two separate helpers that just take the stuff they need.